### PR TITLE
Add OpenTracing support

### PR DESCRIPTION
**Pull request by alesj**
Introducing OpenTelemetry tracing option.

The new tracing module abstracts away the tracing usage -- either previous OpenTracing or new OpenTelemetry support. Instead of same tracing config code being applied over all test-client modules, we handle this in a single tracing module.
Different tracing implementations are triggered by setting implementation specific service name environment variable (see tracing's README for more details).
Each tracing implementation sets a global Tracer instance, which can then be used in supported libraries to provide new traces / spans.
Additionally to service name env var, other required env vars must be set as well -- see README.
Currently tracing implementations add support for Kafka consumer/producer and Kafka Streams tracing.

Similar implementation is used in Strimzi Operator and Strimzi Bridge,
so it might be useful to reconsider a better place for all this implementation code / artifact.

Signed-off-by: Ales Justin ales.justin@gmail.com

### CODE REVIEW

**Comment by alesj** [comment]
@im-konge can you please check / review this, but not yet merge -- I need to run the full test cycle ...

**Review by im-konge — APPROVED** [review]
Fine by me, assuming it works.

**Comment by alesj** [comment]
OK, full test cycle passing ... ready to merge.

**Comment by im-konge** [comment]
Let's wait for @scholzj or @Frawless for review

**Comment by scholzj** [comment]
@im-konge I think my comments from the previous PR still stand.

**Comment by im-konge** [comment]
@alesj can you please fix the comments which @scholzj  had on previous PR?

File: kafka/producer/src/main/java/ProducerConfiguration.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public static Properties createProperties(ProducerConfiguration config) {
             }
         }
 
-        if ((config.getOauthAccessToken() != null)
+        if (config.getOauthAccessToken() != null
```
Comment (im-konge) [inline kafka/producer/src/main/java/ProducerConfiguration.java:102]: Can you return it to `(config.getOauthAccessToken() != null)` please?

File: kafka/streams/src/main/java/StreamsConfiguration.java [hunk 2]
```diff
@@ -96,8 +96,8 @@ public static Properties createProperties(StreamsConfiguration config) {
             }
         }
 
-        if ((config.getOauthAccessToken() != null)
-            || (config.getOauthTokenEndpointUri() != null && config.getOauthClientId() != null && (config.getOauthRefreshToken() != null || config.getOauthClientSecret() != null)))    {
+        if (config.getOauthAccessToken() != null
+            || config.getOauthTokenEndpointUri() != null && config.getOauthClientId() != null && (config.getOauthRefreshToken() != null || config.getOauthClientSecret() != null))    {
```
Comment (im-konge) [inline kafka/streams/src/main/java/StreamsConfiguration.java:100]: Can you return it to previous state? (with parentheses)

**Review by im-konge — COMMENTED** [review]
Just two nits, that are connected to Jakub's comments

**Review by Frawless — APPROVED** [review]
In case it's working I am fine with that. And when you work-in comments from Jakub.

**Comment by alesj** [comment]
I resolved most of the questions / issues in the old PR, apart from these:

```
docker-images/build-images.sh
Outdated
echo "Installing root pom"
make mvn_build

echo "Building Tracing dependency"
make build --directory=tracing/
```

```
@scholzj on 23 Nov 2021
Normally, the submodule would be added into the SUBDIRS list in the main Makefile. So when you run make ... it would trigger it for the tracing module as well and build it that way.

The README.md suggests to call make all to build the project. But if this is only in the build-images.sh script, it would be never called and I assume it would not work with it. That is at least my read on the build.

 @alesj alesj on 23 Nov 2021
Atm the script iterates over kafka and http subdirs, and builds them.
Whereas tracing module is a seperate module on which they depend on,
so it must be built before.

This build-images.sh is only used in CI, afais.
The plain Make makes use of SUBDIRS, where tracing is listed as well.

If @im-konge or @Frawless have any better ideas, fine by me.
```

and

```
http/pom.xml
                <version>${opentracing-kafka.version}</version>
                <groupId>io.strimzi</groupId>
                <artifactId>test-client-tracing</artifactId>
                <version>${project.version}</version>
```

@scholzj Well, It didn't had test-client-tracing dependency from the same project before. So I think you can remove it here.
@alesj So it should go into each submodule?

File: pom.xml [hunk 3]
```diff
@@ -10,25 +10,31 @@
...
-        <opentracing-kafka.version>0.1.15</opentracing-kafka.version>
-        <jaeger.version>1.1.0</jaeger.version>
         <strimzi-oauth-callback.version>0.8.1</strimzi-oauth-callback.version>
         <vertx-core.version>4.1.0</vertx-core.version>
         <netty.version>4.1.66.Final</netty.version>
         <jackson.version>2.12.4</jackson.version>
         <maven-shade.version>3.2.1</maven-shade.version>
-        <kafka-kubernetes-config-provider.version>0.1.0</kafka-kubernetes-config-provider.version>
+        <kafka-kubernetes-config-provider.version>0.1.1</kafka-kubernetes-config-provider.version>
+        <jackson-core.version>2.12.4</jackson-core.version>
```
Comment (im-konge) [inline pom.xml:31]: Why do you have two `jackson` versions?
```
         <jackson.version>2.12.4</jackson.version>
         ...
          <jackson-core.version>2.12.4</jackson-core.version>
```

Comment (alesj) [reply pom.xml]: Ah, good catch, removed.

File: tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        for (TracingHandle instance : HANDLES) {
+            String serviceName = instance.serviceName();
+            if (serviceName != null) {
+                log.info(
+                    String.format(
+                        "Initializing Jaeger (%s) tracingConfig with service name %s",
+                        instance.type(),
+                        serviceName
+                    )
+                );
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java:40]: Cant it just use:
```
log.info("Initializing Jaeger ({}) tracingConfig with service name {}",
                        instance.type(),
                        serviceName)
```
?

File: tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * Tracing initialization
+ */
+public class TracingUtil {
+    private static final Logger log = LogManager.getLogger(TracingUtil.class);
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java:18]: We would normally use uppercase for final static constants. You seem to use it for `HANDLES` as well.

@im-konge This is ideally something what checkstyle should check.

File: tracing/src/main/java/io/strimzi/test/tracing/OpenTracingHandle.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Properties;
+
+public class OpenTracingHandle implements TracingHandle {
+    @Override
+    public String type() {
+        return "OpenTracing";
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/OpenTracingHandle.java:30]: Define the string as `static final` constant.

File: tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    @Override
+    public String envName() {
+        return "OTEL_SERVICE_NAME";
+    }
+
+    @Override
+    public String serviceName() {
+        String serviceName = System.getenv(envName());
+        if (serviceName == null) {
+            serviceName = System.getProperty("otel.service.name");
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java:50]: Define the string as `static final` constant.

File: tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+public class OpenTelemetryHandle implements TracingHandle {
+    @Override
+    public String type() {
+        return "OpenTelemetry";
+    }
+
+    @Override
+    public String envName() {
+        return "OTEL_SERVICE_NAME";
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java:43]: Define the string as `static final` constant.

File: tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    @Override
+    public String serviceName() {
+        String serviceName = System.getenv(envName());
+        if (serviceName == null) {
+            serviceName = System.getProperty("otel.service.name");
+        } else {
+            System.setProperty("otel.service.name", serviceName);
+        }
+        if (serviceName != null && System.getenv("OTEL_TRACES_EXPORTER") == null) {
+            System.setProperty("otel.traces.exporter", "jaeger");
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java:55]: Define the string as `static final` constant.

File: tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Map;
+import java.util.Properties;
+
+public class OpenTelemetryHandle implements TracingHandle {
+    @Override
+    public String type() {
+        return "OpenTelemetry";
```
Comment (scholzj) [inline tracing/src/main/java/io/strimzi/test/tracing/OpenTelemetryHandle.java:38]: Define the string as `static final` constant.

File: kafka/producer/src/main/java/Main.java [hunk 11]
```diff
@@ -26,20 +22,15 @@
 public class Main {
     private static final Logger LOGGER = LogManager.getLogger(Main.class);
 
-    public static void main(String[] args) throws InterruptedException {
+    public static void main(String[] args) throws Exception {
```
Comment (scholzj) [inline kafka/producer/src/main/java/Main.java:25]: Why did the exception it throws changed?

File: tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java (hunk 5 above)
Comment (im-konge) [reply tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java:18]: Sure, will add it there. Good catch

File: tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java (hunk 4 above)
Comment (alesj) [reply tracing/src/main/java/io/strimzi/test/tracing/TracingUtil.java:40]: Fixed.

**Review by scholzj — APPROVED** [review]
LGTM assuming it works.

**Pull request merged**